### PR TITLE
One keyring item for session

### PR DESCRIPTION
**Pull request by nickatsegment**
Fixes: #146

Partially addresses #173 

Adds a `--session-cache-single-item` flag and env var to opt into a new session cache that stores all sessions in a single keyring item. The benefit is fewer auth prompts from macOS Keychain. On every upgrade (new aws-okta binary), the user needs to reauth 3 + N (number of sessions) times, but with this new cache, this is always just 3+1.

This refactors the `KeyringSessions` type into a session cache store and a keying function interface so that we can reuse the existing (bizarre) keying function with a new storage engine.

I might propose in another PR that we use a better keying function. I'm still trying to understand what we want from this cache.

### CODE REVIEW

File: lib/provider.go [hunk 1]
```diff
@@ -91,44 +111,66 @@ func (p *Provider) Retrieve() (credentials.Value, error) {
...
+		ProfileName: source,
+		ProfileConf: profileConf,
+		Duration:    p.SessionDuration,
+	}
+
+	var cachedSession *sessioncache.Session
+	cachedSession, err := p.sessions.Get(key)
 	if err != nil {
-		session, err = p.getSamlSessionCreds()
+		creds, err := p.getSamlSessionCreds()
```
Comment (systemizer) [inline lib/provider.go]: small nit: using the same variable name in this scope and the outer scope _might_ confuse a future engineer

File: lib/provider.go [hunk 2]
```diff
@@ -59,13 +64,18 @@ func (o ProviderOptions) ApplyDefaults() ProviderOptions {
 	return o
 }
 
+type SessionCacheInterface interface {
```
Comment (systemizer) [inline lib/provider.go]: awesome 👐

File: internal/sessioncache/key_orig.go [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+// Key returns a key for the keyring item. This is a string containing the source profile name,
+// the profile name, and a hash of the duration
+//
+// this is a copy of KeyringSessions.key and should preserve behavior, *except* that it assumes `profileName`
+// is a valid and existing profile name
+func (k OrigKey) Key() string {
+	// nick: I don't understand this at all. This key function is roughly:
+	// sourceProfileName + hex(md5(duration + json(profileConf)))
+	// - why md5?
+	// - why the JSON of the whole profile? (especially strange considering JSON map order is undetermined)
```
Comment (Fauzyy) [inline internal/sessioncache/key_orig.go]: Do we also want another implementation of the `Key` interface in this PR to play around with, or maybe that should come later

**Comment by nickatsegment** [comment]
I'm gonna add some simple tests.

File: internal/sessioncache/key_orig.go (hunk 3 above)
Comment (nickatsegment) [reply internal/sessioncache/key_orig.go]: In the new tests, I've got a thing called `fixedKey` that just returns the same string.

If you're thinking about actual, potentially useful ones, I think we'd need to understand what this cache is supposed to accomplish first ;)

**Pull request merged**